### PR TITLE
Replace tylerb/graceful with inbuilt methods Go 1.8 onwards

### PR DESCRIPTION
**Pull request by tortuoise**
For simplicity, examples not re-generated in this commit.

### CODE REVIEW

File: generator/templates/server/server.gotmpl [hunk 1]
```diff
@@ -562,3 +596,37 @@ func (s *Server) TLSListener() (net.Listener, error) {
 	}
 	return s.httpsServerL, nil
 }
+
+func (s *Server) interruptChan() chan os.Signal {
```
Comment (casualjim) [inline generator/templates/server/server.gotmpl]: why is this lazily initialized and not wihen everything else in intialized?
initializing it early removes the need for the mutex

File: generator/templates/server/server.gotmpl [hunk 2]
```diff
@@ -562,3 +596,37 @@ func (s *Server) TLSListener() (net.Listener, error) {
...
+
+}
+
+func (s *Server) handleInterrupt(interrupt chan os.Signal, quitting chan struct{}, l net.Listener) {
+    for _ = range interrupt {
+        if s.interrupted {
+            s.Logf("Server already shutting down")
+            continue
+        }
+        s.interrupted = true
```
Comment (casualjim) [inline generator/templates/server/server.gotmpl]: these 4 would be better if they used an atomic.
This entire method can be run in a sync.Once actually

File: generator/templates/server/server.gotmpl (hunk 1 above)
Comment (tortuoise) [reply generator/templates/server/server.gotmpl]: Ok agreed.

File: generator/templates/server/server.gotmpl (hunk 2 above)
Comment (tortuoise) [reply generator/templates/server/server.gotmpl]: Only niladic functions can run in sync.Once - and I don't think they can have receivers, i.e. be methods.

Comment (casualjim) [reply generator/templates/server/server.gotmpl]: it's just a wrapper around the entire block of code, you have to lift the variables you want to capture out of there by declaring those outside of the closure

**Comment by casualjim** [comment]
I see what you meant. Why is there an interrupt handler for each listener? wouldn't there be 1 interrupt signal that needs to clean up all the active listeners?

File: generator/templates/server/server.gotmpl [hunk 3]
```diff
@@ -331,24 +344,26 @@ func (s *Server) Serve() (err error) {
...
 
 	if s.hasScheme(schemeHTTPS) {
-		httpsServer := &graceful.Server{Server: new(http.Server)}
+		httpsServer := new(http.Server)
 		httpsServer.MaxHeaderBytes = int(s.MaxHeaderSize)
 		httpsServer.ReadTimeout = s.TLSReadTimeout
 		httpsServer.WriteTimeout = s.TLSWriteTimeout
 		httpsServer.SetKeepAlivesEnabled(int64(s.TLSKeepAlive) > 0)
-		httpsServer.TCPKeepAlive = s.TLSKeepAlive
+		//httpsServer.TCPKeepAlive = s.TLSKeepAlive
```
Comment (casualjim) [inline generator/templates/server/server.gotmpl]: please remove commented out item

File: generator/templates/server/server.gotmpl [hunk 4]
```diff
@@ -303,22 +316,22 @@ func (s *Server) Serve() (err error) {
...
 
 	if s.hasScheme(schemeHTTP) {
-		httpServer := &graceful.Server{Server: new(http.Server)}
+		httpServer := new(http.Server)
 		httpServer.MaxHeaderBytes = int(s.MaxHeaderSize)
 		httpServer.ReadTimeout = s.ReadTimeout
 		httpServer.WriteTimeout = s.WriteTimeout
 		httpServer.SetKeepAlivesEnabled(int64(s.KeepAlive) > 0)
-		httpServer.TCPKeepAlive = s.KeepAlive
+		//httpServer.TCPKeepAlive = s.KeepAlive
```
Comment (casualjim) [inline generator/templates/server/server.gotmpl]: please remove commented line

File: generator/templates/server/server.gotmpl [hunk 5]
```diff
@@ -303,22 +316,22 @@ func (s *Server) Serve() (err error) {
...
 		}
 
 		if int64(s.CleanupTimeout) > 0 {
-			httpServer.Timeout = s.CleanupTimeout
+			httpServer.IdleTimeout = s.CleanupTimeout
 		}
 
 		httpServer.Handler = s.handler
-		httpServer.LogFunc = s.Logf
+		//httpServer.LogFunc = s.Logf
```
Comment (casualjim) [inline generator/templates/server/server.gotmpl]: please remove commented line

File: generator/templates/server/server.gotmpl [hunk 6]
```diff
@@ -331,24 +344,26 @@ func (s *Server) Serve() (err error) {
...
-			httpsServer.ListenLimit = s.TLSListenLimit
+			s.httpsServerL = netutil.LimitListener(s.httpsServerL, s.TLSListenLimit)
 		}
 		if int64(s.CleanupTimeout) > 0 {
-			httpsServer.Timeout = s.CleanupTimeout
+			httpsServer.IdleTimeout = s.CleanupTimeout
 		}
 		httpsServer.Handler = s.handler
-		httpsServer.LogFunc = s.Logf
+		//httpsServer.LogFunc = s.Logf
```
Comment (casualjim) [inline generator/templates/server/server.gotmpl]: please remove commented line

File: generator/templates/server/server.gotmpl [hunk 7]
```diff
@@ -562,3 +597,33 @@ func (s *Server) TLSListener() (net.Listener, error) {
...
+	for _ = range s.interrupt {
+		if s.interrupted {
+			s.Logf("Server already shutting down")
+			continue
+		}
+		s.interrupted = true
+		s.Logf("Shutting down... ")
+		close(quitting)
+
+		//s.SetKeepAlivesEnabled(false)
```
Comment (casualjim) [inline generator/templates/server/server.gotmpl]: please remove commented line

File: generator/templates/server/server.gotmpl [hunk 8]
```diff
@@ -562,3 +597,33 @@ func (s *Server) TLSListener() (net.Listener, error) {
...
+		if err := s.httpServerL.Close(); err != nil {
+			s.Logf("Error: %s", err)
+		}
+		if err := s.httpsServerL.Close(); err != nil {
+			s.Logf("Error: %s", err)
+		}
+		if err := s.domainSocketL.Close(); err != nil {
+			s.Logf("Error: %s", err)
+		}
+		//s.shutdown <- struct{}{}
```
Comment (casualjim) [inline generator/templates/server/server.gotmpl]: please remove commented line

Comment (tortuoise) [reply generator/templates/server/server.gotmpl]: Sorry for the stray commented lines - will remove them and re-run generator.

**Pull request closed without merging**